amaranth/hdl/_ast.py

```diff
@@ -399,222 +452,615 @@ def __init__(self, *, src_loc_at=0):
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(max(len(self) - amount, 0))`
+            If :pc:`self` is unsigned.
+        :class:`Value`, :pc:`signed(len(self) - amount)`
+            If :pc:`self` is signed.
+        """
+        if not isinstance(amount, int):
+            raise TypeError(f"Shift amount must be an integer, not {amount!r}")
+        if amount < 0:
+            return self.shift_left(-amount)
+        if self.shape().signed:
+            return self[amount:].as_signed()
+        else:
+            return self[amount:] # unsigned
+
+    def rotate_right(self, amount):
+        """Right rotate by constant amount.
+
+        If :pc:`amount < 0`, performs the same operation as :pc:`self.rotate_right(-amount)`.
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(len(self))`, :ref:`assignable <lang-assignable>`
+        """
+        if not isinstance(amount, int):
+            raise TypeError(f"Rotate amount must be an integer, not {amount!r}")
+        if len(self) != 0:
+            amount %= len(self)
+        return Cat(self[amount:], self[:amount])
+
+    def __len__(self):
+        """Bit width of :pc:`self`.
+
+        Returns
+        -------
+        :class:`int`
+            :pc:`self.shape().width`
+        """
+        return self.shape().width
+
+    def __getitem__(self, key):
+        """Bit slicing.
+
+        .. todo::
+
+            Describe this operation.
+        """
+        n = len(self)
+        if isinstance(key, int):
+            if key not in range(-n, n):
+                raise IndexError(f"Index {key} is out of bounds for a {n}-bit value")
+            if key < 0:
+                key += n
+            return Slice(self, key, key + 1, src_loc_at=1)
+        elif isinstance(key, slice):
+            if isinstance(key.start, Value) or isinstance(key.stop, Value):
+                raise TypeError(f"Cannot slice value with a value; use Value.bit_select() or "
+                                f"Value.word_select() instead")
+            start, stop, step = key.indices(n)
+            if step != 1:
+                return Cat(self[i] for i in range(start, stop, step))
+            return Slice(self, start, stop, src_loc_at=1)
+        elif isinstance(key, Value):
+            raise TypeError(f"Cannot index value with a value; use Value.bit_select() instead")
+        else:
+            raise TypeError(f"Cannot index value with {key!r}")
 
-        Returns
-        -------
-        Value, out
-            ``1`` if an odd number of bits are set, ``0`` if an even number of bits are set.
-        """
-        return Operator("r^", [self])
+    def __contains__(self, other):
+        """Forbidden membership test operator.
 
-    def implies(premise, conclusion):
-        """Implication.
+        Python requires this operator to return a :class:`bool`. Since this is not possible
+        for Amaranth values, this operator always raises an exception.
```

Comment:
We could add a note (here and/or in the error message?) that `.matches()` provides equivalent functionality (and then some) if you want to check membership in a constant set.

Reply:
Oh, good catch!